### PR TITLE
added pylint config to run multiple threads for pylint

### PR DESCRIPTION
**Pull request by gsnider2195**
# Closes: #DNE
# What's Changed

From https://pylint.readthedocs.io/en/stable/user_guide/usage/run.html#parallel-execution

> It is possible to speed up the execution of Pylint. If the running computer has more CPUs than one, then the work for checking all files could be spread across all cores via Pylints's sub-processes. This functionality is exposed via the -j command-line parameter. If the provided number is 0, then the total number of CPUs will be autodetected and used.

On my machine (6C6T) this decreases the run time from 13 mins to 4 mins.

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: changes/4611.changed [hunk 1]
```diff
@@ -0,0 +1 @@
+Updated pylint to use multiple threads.
```
Comment (HanlinMiao) [inline changes/4611.changed:1]: Should this be a `.housekeeping`?

**Review by HanlinMiao — APPROVED** [review]

File: changes/4611.changed (hunk 1 above)
Comment (gsnider2195) [reply changes/4611.changed:1]: I think it definitely could be

**Comment by bryanculver** [comment]
With: `poetry run invoke pylint  1.38s user 0.69s system 1% cpu 1:44.87 total`
Without: `poetry run invoke pylint  2.68s user 1.54s system 1% cpu 5:05.52 total`

:shipit:

**Comment by jathanism** [comment]
Hmm on my machine for some reason, I'm seeing this with `jobs = 0`: 

```
$ time invoke pylint
nautobot-server pylint nautobot tasks.py
03:25:16.266 DEBUG   nautobot.core.celery __init__.py        import_jobs_as_celery_tasks() :
  Importing system Jobs
03:25:16.266 DEBUG   nautobot.core.celery __init__.py                      register_jobs() :
  Registering job nautobot.core.jobs.GitRepositorySync
03:25:16.267 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def process_webhook(webhook_pk, data, model_name, event, timestamp, username, request_id, snapshots):
    return 1

03:25:16.267 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def group(self, tasks, result, group_id, partial_args, add_to_parent=0):
    return 1

03:25:16.268 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def provision_field(field_id, content_type_pk_set):
    return 1

03:25:16.268 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def xmap(task, it):
    return 1

03:25:16.268 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def chain(*args, **kwargs):
    return 1

03:25:16.268 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def backend_cleanup():
    return 1

03:25:16.268 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def delete_custom_field_data(field_key, content_type_pk_set):
    return 1

03:25:16.268 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def xstarmap(task, it):
    return 1

03:25:16.268 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def update_custom_field_choice_data(field_id, old_value, new_value):
    return 1

03:25:16.268 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def chord(self, header, body, partial_args=0, interval=1, countdown=2, max_retries=3, eager=4, **kwargs):
    return 1

03:25:16.268 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def accumulate(self, *args, **kwargs):
    return 1

03:25:16.269 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def get_releases(pre_releases=0):
    return 1

03:25:16.269 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def chunks(task, it, n):
    return 1

03:25:16.269 DEBUG   celery.utils.functional functional.py                    head_from_fun() :

def unlock_chord(self, group_id, callback, interval=0, max_retries=1, result=2, Result=3, GroupResult=4, result_from_tuple=5, **kwargs):
    return 1

03:25:16.269 DEBUG   nautobot.core.celery __init__.py                      register_jobs() :
  Registering job nautobot.core.jobs.GitRepositoryDryRun
03:25:16.269 DEBUG   nautobot.core.celery __init__.py        import_jobs_as_celery_tasks() :
  Importing Jobs from hidden in JOBS_ROOT
03:25:16.270 DEBUG   nautobot.core.celery __init__.py        import_jobs_as_celery_tasks() :
  Importing Jobs from hooks in JOBS_ROOT
03:25:16.270 DEBUG   nautobot.core.celery __init__.py        import_jobs_as_celery_tasks() :
  Importing Jobs from testjob in JOBS_ROOT
03:25:16.642 ERROR   nautobot.core.apps   __init__.py                                url() :
  Error in link construction for Notes: Reverse for 'note_list' not found. 'note_list' is not a valid view function or pattern name.
03:25:16.732 DEBUG   nautobot.core.celery __init__.py                      register_jobs() :
  Registering job example_plugin.jobs.ExampleDryRunJob
03:25:16.732 DEBUG   nautobot.core.celery __init__.py                      register_jobs() :
  Registering job example_plugin.jobs.ExampleJob
03:25:16.732 DEBUG   nautobot.core.celery __init__.py                      register_jobs() :
  Registering job example_plugin.jobs.ExampleHiddenJob
03:25:16.733 DEBUG   nautobot.core.celery __init__.py                      register_jobs() :
  Registering job example_plugin.jobs.ExampleLoggingJob
03:25:16.733 DEBUG   nautobot.core.celery __init__.py                      register_jobs() :
  Registering job example_plugin.jobs.ExampleJobHookReceiver
03:25:16.733 DEBUG   nautobot.core.celery __init__.py                      register_jobs() :
  Registering job example_plugin.jobs.ExampleSimpleJobButtonReceiver
03:25:16.733 DEBUG   nautobot.core.celery __init__.py                      register_jobs() :
  Registering job example_plugin.jobs.ExampleComplexJobButtonReceiver
Using config file /Users/jathan/sandbox/src/nautobot/pyproject.toml
concurrent.futures.process._RemoteTraceback:
"""
Traceback (most recent call last):
  File "/Users/jathan/.pyenv/versions/3.10.6/lib/python3.10/concurrent/futures/process.py", line 246, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/Users/jathan/.pyenv/versions/3.10.6/lib/python3.10/concurrent/futures/process.py", line 205, in _process_chunk
    return [fn(*args) for args in chunk]
  File "/Users/jathan/.pyenv/versions/3.10.6/lib/python3.10/concurrent/futures/process.py", line 205, in <listcomp>
    return [fn(*args) for args in chunk]
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/pylint/lint/parallel.py", line 76, in _worker_check_single_file
    _worker_linter.check_single_file_item(file_item)
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 762, in check_single_file_item
    with self._astroid_module_checker() as check_astroid_module:
  File "/Users/jathan/.pyenv/versions/3.10.6/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 1015, in _astroid_module_checker
    checker.open()
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/pylint_django/checkers/foreign_key_strings.py", line 92, in open
    django.setup()
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/django/__init__.py", line 19, in setup
    configure_logging(settings.LOGGING_CONFIG, settings.LOGGING)
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/django/conf/__init__.py", line 82, in __getattr__
    self._setup(name)
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/django/conf/__init__.py", line 69, in _setup
    self._wrapped = Settings(settings_module)
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/django/conf/__init__.py", line 170, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
  File "/Users/jathan/.pyenv/versions/3.10.6/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1004, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'nautobot_config'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/bin/nautobot-server", line 6, in <module>
    sys.exit(main())
  File "/Users/jathan/sandbox/src/nautobot/nautobot/core/cli/__init__.py", line 54, in main
    run_app(
  File "/Users/jathan/sandbox/src/nautobot/nautobot/core/runner/runner.py", line 297, in run_app
    management.execute_from_command_line([runner_name, command] + command_args)
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/Users/jathan/sandbox/src/nautobot/examples/example_plugin/example_plugin/management/commands/pylint.py", line 23, in handle
    run_pylint(args)
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/pylint/__init__.py", line 36, in run_pylint
    PylintRun(argv or sys.argv[1:])
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/pylint/lint/run.py", line 215, in __init__
    linter.check(args)
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 693, in check
    check_parallel(
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.10/lib/python3.10/site-packages/pylint/lint/parallel.py", line 158, in check_parallel
    for (
  File "/Users/jathan/.pyenv/versions/3.10.6/lib/python3.10/concurrent/futures/process.py", line 570, in _chain_from_iterable_of_lists
    for element in iterable:
  File "/Users/jathan/.pyenv/versions/3.10.6/lib/python3.10/concurrent/futures/_base.py", line 621, in result_iterator
    yield _result_or_cancel(fs.pop())
  File "/Users/jathan/.pyenv/versions/3.10.6/lib/python3.10/concurrent/futures/_base.py", line 319, in _result_or_cancel
    return fut.result(timeout)
  File "/Users/jathan/.pyenv/versions/3.10.6/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/Users/jathan/.pyenv/versions/3.10.6/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
ModuleNotFoundError: No module named 'nautobot_config'
invoke pylint  8.64s user 0.83s system 139% cpu 6.796 total
```

**Comment by gsnider2195** [comment]
> hidden in JOBS_ROOT

I'm not able to reproduce this. I wonder if one of the jobs in your jobs root is causing this?

**Comment by glennmatthews** [comment]
Runs successfully here in about 7 minutes, haven't timed it out definitively but pretty sure that's a substantial improvement over baseline. Nice find!

**Comment by jathanism** [comment]
> I'm not able to reproduce this. I wonder if one of the jobs in your jobs root is causing this?

I tried nuking my jobs dir. Same result. I want this to work so badly! Hate being the outlier!

I tested it on both of my M1 Macs with the same result. At least that is consistent.  :(

**Comment by gsnider2195** [comment]
@bryanculver can you try to replicate the failure @jathanism is seeing? You should be able to run pylint locally with these commands:

```
git checkout u/gas-pylint-threading
export NAUTOBOT_CONFIG=$PWD/development/nautobot_config.py
poetry install
nautobot-server pylint nautobot tasks.py
```

It ran fine for me locally but I'm on x86 linux. Everybody else is on x86 macos.

**Comment by jathanism** [comment]
> @bryanculver can you try to replicate the failure @jathanism is seeing? You should be able to run pylint locally with these commands:
> 
> ```
> git checkout u/gas-pylint-threading
> export NAUTOBOT_CONFIG=$PWD/development/nautobot_config.py
> poetry install
> nautobot-server pylint nautobot tasks.py
> ```
> 
> It ran fine for me locally but I'm on x86 linux. Everybody else is on x86 macos.

Same issue running it this way. 

I did confirm though that running it in Docker works fine.

```
------------------------------------
Your code has been rated at 10.00/10

invoke pylint  0.99s user 0.40s system 1% cpu 1:38.25 total
```

This seems relevant: https://github.com/pylint-dev/pylint/issues/7938

**Comment by gsnider2195** [comment]
Is your local dev environment running an older version of pylint

**Comment by jathanism** [comment]
Got it working locally with this command:

```
$ time pylint --init-hook "import nautobot; nautobot.setup(); import sys; sys.path.append('/Users/jathan/.nautobot')" --rcfile pyproject.toml nautobot
------------------------------------
Your code has been rated at 10.00/10

pylint --init-hook  --rcfile pyproject.toml nautobot  375.59s user 12.38s system 775% cpu 50.000 total
```

Obviously this isn't a complete solution, but it does prove that it will work. It appears to be an issue related to the configuration not being found in the `PYTHONPATH` when the forking is happening.

I am admittedly confused why the `pylint` management command is only published by the `example_plugin`...

**Comment by gsnider2195** [comment]
I was also perplexed by the management command in the example plugin 😆

**Comment by glennmatthews** [comment]
IIRC the reasoning was that it isn't a command that should be available (as it won't work) in "production" Nautobot installations. There may be a more elegant way to address that though.

**Comment by gsnider2195** [comment]
@jathanism what version of pylint are you encountering that error on?

**Comment by jathanism** [comment]
> @jathanism what version of pylint are you encountering that error on?

Latest version.

**Review by glennmatthews — APPROVED** [review]

**Comment by jathanism** [comment]
> Any chance the workaround mentioned in [pylint-dev/pylint-django#370 (comment)](https://github.com/pylint-dev/pylint-django/issues/370#issuecomment-1305725496) (explicitly setting PYTHONPATH) might address your issue, @jathanism?

If you look at the thing that worked for me, that's expliciltly what I had to do, however, it has to be done *for each thread* not globally. I did try explicitly setting `PYTHONPATH=.`, `PYTHONPATH=base_dirpath/to/nautobot_config.py`, and various iterations on that them. This is the only thing that got it to work for me:

```
pylint --init-hook "import nautobot; nautobot.setup(); import sys; sys.path.append('/Users/jathan/.nautobot')" --rcfile pyproject.toml nautobot
```

**Pull request merged**